qa/tasks/cephfs/test_scrub.py

```diff
@@ -176,3 +176,12 @@ def test_scrub_backtrace(self):
 
     def test_scrub_dup_inode(self):
         self._scrub(DupInodeWorkload(self, self.fs, self.mount_a))
+
+    def test_mdsdir_scrub_backtrace_repair(self):
+        damage_count = self._get_damage_count()
+
+        self.fs.run_scrub(["start", "~mdsdir", "recursive"])       
+        self.assertEqual(self._get_damage_count(), damage_count + 1)        
```

Comment:
Is it always guaranteed to find a damaged inode on recursive scrub ?

Reply:
As mentioned in the tracker and I also did get it every time after creation of cluster when I do recursive scrub, which is removed after we `rm` it.

Reply:
> As mentioned in the tracker and I also did get it every time after creation of cluster when I do recursive scrub, which is removed after we `rm` it. And couldn't be resolved with `recursive repair` earlier.

Reply:
And the test succeeds, every number of time I ran the tests.

Reply:
@neesingh-rh could you paste the damage here ?
it might be a bug lurking around e.g. missed mdlog flush

damage getting reported on a freshly created filesystem/cluster doesn't sound good
does this happen on a vstart cluster as well ?

Reply:
What? Scrub finds a backtrace damage on a _just_ created file system?

Reply:
> @neesingh-rh could you paste the damage here ? it might be a bug lurking around e.g. missed mdlog flush
> 
```[neerajpratapsingh@fedora build]$ ./bin/ceph tell mds.0 damage ls
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2022-10-28T12:03:11.160+0530 7f0d71c72640 -1 WARNING: all dangerous and experimental features are enabled.
2022-10-28T12:03:11.181+0530 7f0d70a10640 -1 WARNING: all dangerous and experimental features are enabled.
2022-10-28T12:03:11.203+0530 7f0d427fc640  0 client.4166 ms_handle_reset on v1:192.168.1.7:6815/3230467377
2022-10-28T12:03:11.242+0530 7f0d427fc640  0 client.4168 ms_handle_reset on v1:192.168.1.7:6815/3230467377
[
    {
        "damage_type": "backtrace",
        "id": 143312872,
        "ino": 256,
        "path": "~mds0"
    }
]```
> damage getting reported on a freshly created filesystem/cluster doesn't sound good does this happen on a vstart cluster as well ?
>
Yeah, if I start vstart cluster and do a scrub on `~mdsdir` , I could see this damage.

Reply:
> damage getting reported on a freshly created filesystem/cluster doesn't sound good does this happen on a vstart cluster as well ?
>
Yeah, if I start vstart cluster and do a scrub on `~mdsdir` , I could see this damage.

Reply:
@neesingh-rh Did you debug why backtrace damage is reported on ~mdsdir? We would want to fix that too...

Reply:
I see this in MDS log:

`
2022-10-31T13:38:41.354+0530 7ff8858b3640 -1 mds.0.scrubstack _validate_inode_done scrub error on inode [inode 0x100 [...2,head] ~mds0/ auth v1 ap=1 snaprealm=0x55d665aa5400 DIRTYPARENT f(v0 10=0+10) n(v0 rc2022-10-31T13:38:12.570974+0530 11=0+11) (inest lock) (iversion lock) | lock=0 dirfrag=1 dirtyparent=1 authpin=1 scrubqueue=0 0x55d660d47180]: {"performed_validation":true,"passed_validation":false,"backtrace":{"checked":true,"passed":false,"read_ret_val":-61,"ondisk_value":"(-1)0x0:[]//[]","memoryvalue":"(2)0x100:[]//[]","error_str":"failed to read off disk; see retval"},"raw_stats":{"checked":true,"passed":true,"read_ret_val":0,"ondisk_value.dirstat":"f(v0 10=0+10)","ondisk_value.rstat":"n(v0 rc2022-10-31T13:38:12.570974+0530 11=0+11)","memory_value.dirstat":"f(v0 10=0+10)","memory_value.rstat":"n(v0 rc2022-10-31T13:38:12.570974+0530 11=0+11)","error_str":""},"return_code":-61}
`

@vshankar

Reply:
> I see this in MDS log:
> 
> `2022-10-31T13:38:41.354+0530 7ff8858b3640 -1 mds.0.scrubstack _validate_inode_done scrub error on inode [inode 0x100 [...2,head] ~mds0/ auth v1 ap=1 snaprealm=0x55d665aa5400 DIRTYPARENT f(v0 10=0+10) n(v0 rc2022-10-31T13:38:12.570974+0530 11=0+11) (inest lock) (iversion lock) | lock=0 dirfrag=1 dirtyparent=1 authpin=1 scrubqueue=0 0x55d660d47180]: {"performed_validation":true,"passed_validation":false,"backtrace":{"checked":true,"passed":false,"read_ret_val":-61,"ondisk_value":"(-1)0x0:[]//[]","memoryvalue":"(2)0x100:[]//[]","error_str":"failed to read off disk; see retval"},"raw_stats":{"checked":true,"passed":true,"read_ret_val":0,"ondisk_value.dirstat":"f(v0 10=0+10)","ondisk_value.rstat":"n(v0 rc2022-10-31T13:38:12.570974+0530 11=0+11)","memory_value.dirstat":"f(v0 10=0+10)","memory_value.rstat":"n(v0 rc2022-10-31T13:38:12.570974+0530 11=0+11)","error_str":""},"return_code":-61}`
> 
> @vshankar

Yeah, it is pasted in the tracker too. I am trying to figure out the reason for these failure codes.

Reply:
~mdsdir probably does not have an inode backtrace (errno: -61/ENODATA). You might want to check if its required to if it can be skipped. HTH.

Reply:
Okay, for this we still have to find the reason for this damage, right?

Reply:
> Okay, for this we still have to find the reason for this damage.

It's what Venky said — the ~mdsdir didn't write a backtrace to disk at all.

I thought that was a state that didn't trigger damage normally, though — I'm not sure if the mdsdir is special in some way which breaks the safety check or what. Maybe because the mdsdir's live version is also 0 (perhaps it never updates?) and so scrub doesn't recognize the on disk state as being stale?

Reply:
@neesingh-rh At the very bottom of `void MDCache::create_mydir_hierarchy(MDSGather *gather)`, adding the following two lines should do the trick:
```
        my->mark_dirty_parent(ls, true);
        my->store_backtrace(gather->new_sub());
```

Reply:
or maybe there's no persistent dentry pertaining to those **~mdsdir** inodes

Reply:
> or maybe there's no persistent dentry pertaining to those **~mdsdir** inodes

stray dirs are under 0x100 (100.00000000 object). Its just that 0x100 does not have layout or parent xattr.